enhancements/coreos-bootimages.md

```diff
@@ -0,0 +1,143 @@
+
+This enhancement calls for a standardized JSON format for (RHEL) CoreOS bootimage metadata to be placed at https://mirror.openshift.com *and* included in a new `rhel-coreos-bootimages` image included in the release image.
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Since the initial release of OpenShift 4, we have "pinned" RHCOS bootimage metadata inside [openshift/installer](https://github.com/openshift/installer).
+In combination with the binding between the installer and release image, this means that everything needed to install OpenShift (including the operating system "bootimages" such as e.g. AMIs and OpenStack `.qcow2` files) are all captured behind the release image which we can test and ship as an atomic unit.
+
+We have a mechanism to do [in place updates](https://github.com/openshift/machine-config-operator/blob/master/docs/OSUpgrades.md), but there is no automated mechanism to update "bootimages" past a cluster installation.
+
+This enhancement does not describe an automated mechanism to do this: the initial goal is to include this metadata in a standardized format in the cluster and at mirror.openshift.com so that UPI installations can do this manually, and we can start work on an IPI mechanism.
+
+### Stream metadata format
+
+As part of unifying Fedora CoreOS and RHEL CoreOS, we have standardized on the "stream metadata" format used by FCOS.  More in [FCOS docs](https://docs.fedoraproject.org/en-US/fedora-coreos/getting-started/) and [this RHCOS issue](https://github.com/openshift/os/issues/477).
+
+There is a new [stream-metadata-go](https://github.com/coreos/stream-metadata-go) library to consume this data.
+
+### Adding a new git repository and release image component + configmap with this data
+
+A new git repository https://github.com/openshift/rhel-coreos-bootimages will be created (based on [an existing prototype repository](https://github.com/cgwalters/rhel-coreos-bootimages)).  It will have the stream JSON.
+
+Additionally, this repository will be included in the release image as `rhel-coreos-bootimages` and it will use the CVO to install a `configmap/coreos-bootimages` in the `openshift-machine-config-operator` namespace.
+
+### Update `oc adm release new` to patch the installer binary to inject this data
+
+The installer needs CoreOS to start at all; today the bootimages are used for the bootstrap node
+and control plane which are created by the installer directly.  In order to avoid duplicating the
+metadata, change `oc adm release new` to *binary patch* the included `openshift-install`
+binary, in the same way we [do for the release image](https://github.com/openshift/installer/pull/1422).
```

Comment:
Hrm.  Patching the installer (like we patch the CLI) is something that is a requirement in our process already, so we wouldn't be worsening it.  It's a bit wierd because now to extract the installer we have to extract something from an image BEFORE we extract the installer.  The JSON blob here is also significantly larger than a version string, so there is some ugliness there (the way we do patching has some limitations, but probably wouldn't block this change).

One alternative is having the installer fetch the metadata the same way the oc tools do, reusing our libaries (which is no less safe, but does change a bit of the installer flow).  That has some advantages because then the installer could verify other properties of the install including whether the credentials the user provides have access to the images, but which would not help when installing into restricted environments where the installer doesn't have access to the things that are pulled at all.  I think given that I'd probably rule out this option.

A second alternative is to make the user fetch this data themselves and provide it to the installer.  This breaks the hermetic quality of the release, and loses the integrity check we get when we install from the installer (right now the shasum of the installer is enough to verify the entire install chain).  Because of the loss of integrity, I'd rule out that option to.

A variant of your original option is to have the metadata about where those come from be part of machine-os-content, either as labels on the image, or a file in machine-os-content, which we extract and place in the payload directly.  However, the implication I think from the team is that we would then need to know the final publish locations before building machine-os-content?  That might not be insurmountable as a label, but might be higher cost to the release pipeline, and so a separate image is maybe ok.

The last alternative is to keep this in the installer, but to have the installer inject the content as a config map on the cluster.  I think of all the options I prefer that the most because it requires no additional change to our release tooling, and requires no additional repo, and since this is purely metadata, I'd prefer the data used by the installer to be colocated with the installer binary.

I'd like to hear arguments against my last alternative, because I prefer it over what is described in this enhancement (maybe I missed a reason why we can't do that)

Reply:
> A variant of your original option is to have the metadata about where those come from be part of machine-os-content, either as labels on the image, or a file in machine-os-content, which we extract and place in the payload directly. However, the implication I think from the team is that we would then need to know the final publish locations before building machine-os-content?

Yes, there are very messy aspects around the buildsystem to include this in `machine-os-content`.  But actually another problem is that any time we change `machine-os-content`, that forces clusters to update in-place.  I recently went for example to do a PR to a release branch that added some extra commentary into a systemd unit we ship, but I stopped when I realized it'd pointlessly force a reboot for customers.

A more recent example is that we have a bug in LUKS for 4.7 - this only requires a bootimage update, there's nothing to change for running nodes.

A pathological case here is when we e.g. just have a fix for vSphere and just want to bump that bootimage.

And more generally I want to try harder to (re)split the bootimage-vs-`machine-os-content` in our build system, partly for this reason.  The other reason to do the split is we have the converse problem where e.g. we just want to get a kernel/crio/kubelet fix out and that *doesn't* need new bootimages, but we still make new AMIs for every new kubelet, which is a pointless waste.

Reply:
Now, there's definitely things we could do to mitigate the "every machine-os-content change rolls the whole cluster for everyone" issue; I'd eventually like to include some sort of metadata around update urgency.  There's a huge difference between a kernel CVE update and my new commentary in a systemd unit.

And we could probably teach the MCO to be more intelligent about this - e.g. right now it just compares image sha256, but we could go back to trying to have it pull the inner ostree out of the image and skip an update if that hasn't changed.  (Well, that plus extensions).  It'd mean we'd still end up causing existing clusters to pull the m-o-c to at least one node, and then skip doing anything - wouldn't be too bad.  But it starts to grow nontrivial logic in the MCO and we're near capacity there for nontrivial changes.

Reply:
A somewhat less good (but quite important to us tactically right now) reason is we're in the midst of a transition of trying to have our CI and build be more Prow native, - but getting `machine-os-content` to work that way all at once is unlikely to happen for 4.8.  But with a new component we can add this new git repository now that would have (just like openshift/installer does) Prow jobs that gate it and we'd suddenly gain for free a ton of OpenShift CI integration that doesn't exist today.

Reply:
> The last alternative is to keep this in the installer, but to have the installer inject the content as a config map on the cluster. I think of all the options I prefer that the most because it requires no additional change to our release tooling, and requires no additional repo, and since this is purely metadata, I'd prefer the data used by the installer to be colocated with the installer binary.

Oh sorry I missed the nuance of this and spent a lot of time replying to the m-o-c bit but here's what you were really arguing!  Sorry.

OK yes, in fact that's *exactly* what I was working on in https://github.com/openshift/installer/pull/4582 but the reason I pivoted to this is because it gets messy to go from there to solving the "how do UPI installs consume this" problem.  Would we add something like `openshift-install print-coreos-stream-metadata` that dumps the JSON? 

Should UPI (including our automated UPI tests) fetch the metadata from `https://mirror.openshift.com` instead?

We're kind of here because https://github.com/openshift/installer/pull/2092 was rejected and the installer team lead at the time was strongly of the opinion this should be part of the release image and `oc`.

Reply:
For MCO,  m-o-c is the source of truth today. About teaching MCO to skip non-trivial OS update, I think one way would be to add a label (like update:true) to m-o-c container that MCD reads and considers as no-op.

Reply:
Clayton and I had a fruitful discussion around his last alternative. I am in agreement with him in preference of that alternative. 

A configmap containing the metadata stream would be added as a manifest in the installer image. This configmap would be injected into the cluster in the normal way for image manifests during installs and upgrades.

There would be a command added to `openshift-install` that a user could run to get the stream metadata, for use cases such as UPI installs.

Reply:
Also of note in the details of the last alternative is that the installer would include the stream metadata configmap in its binary so that the stream metadata used at install-time matches what is included in the release image.